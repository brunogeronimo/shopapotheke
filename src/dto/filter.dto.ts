import { IsArray, IsDate, IsDateString, IsEnum, IsIn, IsISO8601, IsOptional, IsString } from 'class-validator';
import { Limit } from '../enums/limit.enum';
import { Sort } from '../enums/sort.enum';
import { ApiProperty } from '@nestjs/swagger';

export class FilterDto {
  @ApiProperty({
    name: 'languages[]',
    required: false,
    isArray: true,
    type: 'string',
    description: 'Filter by programming language'
  })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  languages: string[];

  @ApiProperty({ required: false, default: '10' })
  @IsOptional()
  @IsEnum(Limit)
  limit: Limit;

  @ApiProperty({ required: false, default: Sort.STARS_DESC })
  @IsOptional()
  @IsEnum(Sort)
  sortBy: Sort;

  @ApiProperty({ required: false })
  @IsOptional()
  @IsString()
  text: string;

  @ApiProperty({ required: false, description: 'Format: YYYY-MM-DD' })
  @IsOptional()
  @IsISO8601()
  createdAfter: string;

  @ApiProperty({ required: false, description: 'Format: YYYY-MM-DD' })
  @IsOptional()
  @IsISO8601()
  createdBefore: string;
}