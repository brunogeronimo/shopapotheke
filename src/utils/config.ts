export const GITHUB_API = process.env.GITHUB_API ?? 'https://api.github.com';
export const GITHUB_API_VERSION = process.env.GITHUB_API_VERSION ?? 'REST';
export const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
export const PORT = process.env.PORT ?? 3000;
