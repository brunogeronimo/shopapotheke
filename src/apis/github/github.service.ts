import { GITHUB_API_VERSION, GITHUB_TOKEN } from '../../utils/config';
import { GraphqlService } from './graphql.service';
import { RestService } from './rest.service';
import { GithubService } from './interfaces/github-service.interface';

export const GITHUB_SERVICE = 'GITHUB_SERVICE';

export function githubService(): GithubService {
  if (GITHUB_API_VERSION === 'GRAPHQL' && !!GITHUB_TOKEN) {
    return new GraphqlService(GITHUB_TOKEN);
  }
  return new RestService();
}