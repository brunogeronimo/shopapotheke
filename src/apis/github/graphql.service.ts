import { Injectable } from '@nestjs/common';
import { GITHUB_API } from '../../utils/config';
import axios from 'axios';
import { FilterDto } from '../../dto/filter.dto';
import { GithubService } from './interfaces/github-service.interface';
import { Repository } from '../../interfaces/repository.interface';
import { Sort } from '../../enums/sort.enum';
import { Limit } from '../../enums/limit.enum';
import { DateFilter } from './enums/date-filter.enum';

@Injectable()
export class GraphqlService implements GithubService {
  constructor(private readonly token: string) {}

  async search(filter: FilterDto): Promise<Repository[]> {
    return this.makeRequest(filter);
  }

  private async makeRequest(filter: FilterDto): Promise<Repository[]> {
    const body = this.prepareBody(filter);

    return new Promise((resolve, reject) => {
      axios.post(`${GITHUB_API}/graphql`,
        {
          query: body
        },
        {
          headers: {
            Authorization: `bearer ${this.token}`
          }
      }).then(response => {
        const github = response.data;

        const repositories: Repository[] = github.data.search.nodes.map(node => {
          return new Repository({
            createdAt: new Date(node.createdAt),
            description: node.description,
            name: node.name,
            owner: node.owner.login,
            stars: node.stargazers.totalCount,
            url: node.url
          });
        });

        resolve(repositories);
      }).catch(error => {
        console.error(error);
        reject(error);
      });
    });
  }

  private prepareBody(filter: FilterDto): string {
    let query = `${filter.text ?? ''} `;

    if (filter.languages) {
      query += `lang:${filter.languages.join(' ')} `;
    }

    if (filter.createdAfter) {
      query += `created:${DateFilter.AFTER}${filter.createdAfter} `;
    }

    if (filter.createdBefore) {
      query += `created:${DateFilter.BEFORE}${filter.createdBefore} `;
    }

    query += `sort:${filter.sortBy ?? Sort.STARS_DESC} `;

    return `
      query {
        search(query: "${query}", type: REPOSITORY, first: ${filter.limit ?? Limit.TEN}) {
          nodes {
            ... on Repository {
              name,
              owner {
                login
              },
              createdAt,
              description,
              url,
              stargazers {
                totalCount
              }
            }
          }
        }
      }
    `;
  }
}