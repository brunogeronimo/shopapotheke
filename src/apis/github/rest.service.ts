import { Injectable } from '@nestjs/common';
import axios from 'axios';
import { GITHUB_API } from '../../utils/config';
import { FilterDto } from '../../dto/filter.dto';
import { GithubService } from './interfaces/github-service.interface';
import { Repository } from '../../interfaces/repository.interface';
import { DateFilter } from './enums/date-filter.enum';
import { Sort } from '../../enums/sort.enum';
import { Limit } from '../../enums/limit.enum';

@Injectable()
export class RestService implements GithubService {
  search(filter: FilterDto): Promise<Repository[]> {
    return this.makeRequest(filter);
  }

  private makeRequest(filter: FilterDto): Promise<Repository[]> {
    return new Promise((resolve, reject) => {
      let query = `${filter.text ?? ''}`;

      if (filter.languages) {
        query += `lang:${filter.languages.join(' ')} `;
      }

      if (filter.createdBefore) {
        query += `created:${DateFilter.BEFORE}${filter.createdBefore} `;
      }

      if (filter.createdAfter) {
        query += `created:${DateFilter.AFTER}${filter.createdAfter} `;
      }

      query += `sort:${filter.sortBy ?? Sort.STARS_DESC}`;

      const params = {
        q: query,
        per_page: filter.limit ?? Limit.TEN
      }

      axios.get(`${GITHUB_API}/search/repositories`, { params })
        .then(response => {
          const github = response.data;

          const repositories: Repository[] = github.items.map(item => {
            return new Repository({
              createdAt: new Date(item.created_at),
              description: item.description,
              name: item.name,
              owner: item.owner.login,
              stars: item.stargazers_count,
              url: item.url
            })
          });

          resolve(repositories);
      }).catch(error => {
        console.error(error);
        reject(error);
      });
    })
  }
}