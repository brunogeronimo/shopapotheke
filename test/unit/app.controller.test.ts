import { Test, TestingModule } from '@nestjs/testing';
import { AppController } from '../../src/app.controller';
import { FilterDto } from '../../src/dto/filter.dto';
import { GITHUB_SERVICE, githubService } from '../../src/apis/github/github.service';
import { RestService } from '../../src/apis/github/rest.service';
import { isEmptyObject } from '../../src/utils/isEmptyObject';
import { Repository } from '../../src/interfaces/repository.interface';

jest.mock('../../src/apis/github/interfaces/github-service.interface');
jest.mock('../../src/apis/github/rest.service');
jest.mock('../../src/utils/isEmptyObject');
jest.mock('../../src/apis/github/github.service');

describe('AppController', () => {
  let appController: AppController;
  let filterDto: FilterDto;
  let restService: RestService;

  beforeEach(async () => {
    filterDto = new FilterDto();
    restService = new RestService();

    const app: TestingModule = await Test.createTestingModule({
      controllers: [AppController],
      providers: [{
        provide: GITHUB_SERVICE,
        useValue: restService
      }],
    }).compile();

    appController = app.get<AppController>(AppController);
  });

  describe('/repositories route', () => {
    it('Should return error if filter is blank"', async (done) => {
      // @ts-ignore
      isEmptyObject.mockImplementation(() => {
        return true;
      });

      appController.getRepositories(filterDto).catch((error) => {
        expect(error).toMatchSnapshot();
        done();
      });
    });

    it('Should return a valid response', async () => {
      // @ts-ignore
      isEmptyObject.mockImplementation(() => {
        return false;
      });
      // @ts-ignore
      githubService.mockImplementation(() => {
        return restService;
      });
      // @ts-ignore
      restService.search = () => {
        return Promise.resolve([
          new Repository({
            createdAt: new Date('2016-09-23T09:34:46.000Z'),
            description: "依赖于 org.common.lang 的 java 工具库",
            name: "tool",
            owner: "OrsonEx",
            stars: 77,
            url: "https://github.com/OrsonEx/tool",
          })
        ]);
      };

      const repositories = await appController.getRepositories(filterDto);

      expect(repositories).toMatchSnapshot();
    })
  });
});
