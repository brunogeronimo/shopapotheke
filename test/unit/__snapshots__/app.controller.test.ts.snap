// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController /repositories route Should return a valid response 1`] = `
Array [
  Repository {
    "createdAt": 2016-09-23T09:34:46.000Z,
    "description": "依赖于 org.common.lang 的 java 工具库",
    "name": "tool",
    "owner": "OrsonEx",
    "stars": 77,
    "url": "https://github.com/OrsonEx/tool",
  },
]
`;

exports[`AppController /repositories route Should return error if filter is blank" 1`] = `[Error: At least one filter has to be set.]`;
